### PR TITLE
RavenDB-18369: Implementation of AndNot as support structure for negatives

### PR DESCRIPTION
**Pull request by redknightlois**
The basis of `AndNot` is that allows us to negate whatever inclusive query we have available.

### CODE REVIEW

**Comment by redknightlois** [comment]
At first I implemented also a `NotInTermProvider` but after implementing the `AndNot` operation all of the uses from the provider get simplified by the use of `AndNot` so much that it doesn't make any sense to deal with the complexities of keeping the hash sets required to avoid the documents that we have already seen. While I have not implemented the SSE/AVX version of it, if we find that it is important, we can do it. 

The `AndNot` operation is much more broader than the implementation of `not in` it can be used in combination with boolean algebra to simplify different types of operation in order to optimize queries when the conditions are are a good match for it.

File: src/Corax/Queries/AndNotMatch.cs [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+
+            _bufferSize = size;
+            _documentBuffer = (long*)buffer.Ptr;
+            _bufferHandler = bufferHandler;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public int AndWith(Span<long> buffer, int matches)
+        {
+            throw new NotSupportedException($"{nameof(AndNotMatch<TInner, TOuter>)} does not support the operation of {nameof(AndWith)}.");
```
Comment (maciejaszyk) [inline src/Corax/Queries/AndNotMatch.cs]: Why?

File: src/Corax/Queries/AndNotMatch.cs [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+            {
+                size = currentLength * 2;
+            }
+
+            // Allocate the new buffer
+            var bufferHandler = _context.Allocate(size * sizeof(long), out var buffer);
+
+            // Ensure we copy the content and then switch the buffers. 
+            new Span<long>(_documentBuffer, currentlyUsed).CopyTo(new Span<long>(buffer.Ptr, size));
+            _bufferHandler.Dispose();
```
Comment (maciejaszyk) [inline src/Corax/Queries/AndNotMatch.cs]: This is wrong. You're returning buffer into the pool, some other thread can take this memory before the query is completed. 

Am I right, @arekpalinski ?

Comment (arekpalinski) [reply src/Corax/Queries/AndNotMatch.cs]: A query is handled by a single thread. So nobody else is supposed to use `_context` that we have here.

Comment (ayende) [reply src/Corax/Queries/AndNotMatch.cs]: Doesn't matter. You are still returning the memory and using it.
What happens if you do another allocation later (that uses the buffer)

Comment (arekpalinski) [reply src/Corax/Queries/AndNotMatch.cs]: We no longer us it. We allocate new buffer here, copy the content from old one and return it. Next allocation can pick it and that won't cause any issue.

Comment (redknightlois) [reply src/Corax/Queries/AndNotMatch.cs]: Unless I am missing something, we are returning the old buffer and keeping the new buffer for use and this is not a buffer that can be exposed to external parties. Check line 199: https://github.com/ravendb/ravendb/pull/13896/files/1b7a5459ab4030bce14ac15c5d81021a7e90a25d#diff-c9c081f143f0ebe0e783389dde9feb524759fc178eb298b917a2b5ebe9a4b24cR199

**Comment by redknightlois** [comment]
@maciejaszyk `AndWith` is now supported. I will build the Memoization block now. Pretty sure it will be ready to be used by Monday.

**Comment by redknightlois** [comment]
Support for memoization. Have 2 examples of usage with `AndNot` and `And`. It ended up being much more involved that I thought, but in yet another tribute to David Wheeler: "All Problems in Computer Science Can Be Solved by Another Layer of Indirection", memoization was not the exception.

**Comment by redknightlois** [comment]
Closing as it will be included and superseded by https://github.com/ravendb/ravendb/pull/13954

**Pull request closed without merging**